File: go/internal/agent/reviewer/publish.go
package reviewer

import (
	"context"
	"fmt"
	"regexp"
	"strings"
	"time"

	"automation-agent/internal/githubapi"
)

// checkName is the advisory check the reviewer publishes (agent-published, human-consumed —
// distinct from the fixers' agent-*-verify checks that the agent reads). Globally unique and
// identical across ports (external contract).
const checkName = "agent-review"

// publishMeta carries the per-PR identifiers and context the published artifacts need.
type publishMeta struct {
	owner, repo string
	number      int
	headSHA     string
	files       []githubapi.PRFile // for the in-diff index
	lenses      []lensStat         // per-lens status rows for the Review details section
	standards   []string           // applied standards source paths (nil = generic), for reporting
}

// summaryMarker is the hidden HTML comment that identifies the reviewer's single summary comment
// so a re-review updates it rather than posting a new one (spec Decision 9).
func summaryMarker(owner, repo string, number int) string {
	return fmt.Sprintf("<!-- automation-agent:review:%s/%s#%d -->", owner, repo, number)
}

// publish posts the review for a scored PR: inline comments for in-diff actionable findings, a
// marker-updated summary comment with the scorecard, and the advisory agent-review check. Out-of-
// diff actionable findings and nitpicks go into the summary (never dropped, spec Decision 6).
func (e *Engine) publish(ctx context.Context, card scorecard, findings []finding, meta publishMeta) error {
	// At-least-once safety: reconciliation makes the inline comments idempotent, but the check run
	// and summary are create/upsert-only, so a redelivered task for a SHA already published would
	// duplicate the check. If the agent-review check already exists for this head SHA, skip — a
	// genuine re-push carries a new SHA and still reconciles below.
	if e.alreadyPublished(ctx, meta) {
		e.log.Info("review already published for head SHA; skipping re-post", "repo", meta.owner+"/"+meta.repo, "sha", meta.headSHA)
		return nil
	}
	idx := newDiffIndex(meta.files)
	inline, outOfDiff, nitpicks := classify(findings, idx)
	actionable := len(inline) + len(outOfDiff)

	// Reconcile against the comments already on the PR (GitHub-as-store): keep inline findings that
	// still apply (don't re-post — so a re-review is idempotent), post only new ones, and minimize
	// the comments whose finding is gone (fixed or no longer relevant). This makes re-publishing on
	// a redelivered task or a re-push safe (spec Decisions 6/11).
	existing, err := e.gh.ListReviewComments(ctx, meta.owner, meta.repo, meta.number)
	if err != nil {
		return fmt.Errorf("reviewer: list existing comments: %w", err)
	}
	rec := reconcile(inline, existing)

	// Post only the new inline findings; an empty review is noise.
	if len(rec.toPost) > 0 {
		comments := make([]githubapi.ReviewComment, 0, len(rec.toPost))
		for _, f := range rec.toPost {
			comments = append(comments, githubapi.ReviewComment{Path: f.File, Line: f.Line, Side: "RIGHT", Body: inlineCommentBody(f)})
		}
		body := fmt.Sprintf("%s Agent review — see the summary comment for the full scorecard.", card.overall)
		// Pin the review to the SHA the findings were computed against. meta.files (and so every
		// line number in comments) came from this SHA's diff; without the pin GitHub would resolve
		// the lines against whatever HEAD is current when this call lands.
		if err := e.gh.CreateReview(ctx, meta.owner, meta.repo, meta.number, githubapi.ReviewInput{Body: body, Comments: comments, CommitID: meta.headSHA}); err != nil {
			return fmt.Errorf("reviewer: post review: %w", err)
		}
	}

	// Minimize the comments whose finding no longer applies — best-effort. New inline comments are
	// already posted but the summary and check are not; aborting here on a single minimize failure
	// would leave the PR without its summary/check, and a retry would short-circuit at
	// alreadyPublished once the check exists. So log and continue per node, letting the summary and
	// check still publish. A leftover stale comment is collapsed on the next genuine re-push.
	for _, id := range rec.toMinimize {
		if err := e.gh.MinimizeComment(ctx, id); err != nil {
			e.log.Warn("reviewer: minimize outdated comment failed; continuing", "repo", meta.owner+"/"+meta.repo, "node", id, "err", err)
		}
	}

	marker := summaryMarker(meta.owner, meta.repo, meta.number)
	if err := e.gh.UpsertMarkerComment(ctx, meta.owner, meta.repo, meta.number, marker, summaryComment(marker, card, actionable, nitpicks, outOfDiff, meta)); err != nil {
		return fmt.Errorf("reviewer: upsert summary comment: %w", err)
	}

	if err := e.gh.CreateCheckRun(ctx, meta.owner, meta.repo, githubapi.CheckRunInput{
		Name:       checkName,
		HeadSHA:    meta.headSHA,
		Conclusion: checkConclusion(card.overall),
		Title:      fmt.Sprintf("%s Agent review — %s", card.overall, levelWord(card.overall)),
		Summary:    fmt.Sprintf("Overall: %s · Actionable comments: %d", levelWord(card.overall), actionable),
	}); err != nil {
		return fmt.Errorf("reviewer: create check: %w", err)
	}
	return nil
}

// publishDeny posts the "too large to review" outcome (spec Decision 4): a marker-updated summary
// comment framed fail-like (🔴) plus a neutral check. No model call was made.
func (e *Engine) publishDeny(ctx context.Context, meta publishMeta, reason string, files, diffBytes int) error {
	if e.alreadyPublished(ctx, meta) {
		e.log.Info("deny already published for head SHA; skipping re-post", "repo", meta.owner+"/"+meta.repo, "sha", meta.headSHA)
		return nil
	}
	marker := summaryMarker(meta.owner, meta.repo, meta.number)
	body := fmt.Sprintf("%s\n## 🔴 Agent review — too large for automated review\n\nThis PR is too large to review automatically (%d files / %d bytes after excluding generated files). Please split it into smaller PRs.\n\n_%s_\n",
		marker, files, diffBytes, reason)
	if err := e.gh.UpsertMarkerComment(ctx, meta.owner, meta.repo, meta.number, marker, body); err != nil {
		return fmt.Errorf("reviewer: upsert deny comment: %w", err)
	}
	if err := e.gh.CreateCheckRun(ctx, meta.owner, meta.repo, githubapi.CheckRunInput{
		Name:       checkName,
		HeadSHA:    meta.headSHA,
		Conclusion: "neutral",
		Title:      "🔴 Agent review — too large",
		Summary:    fmt.Sprintf("%d files / %d bytes after excluding generated files; please split.", files, diffBytes),
	}); err != nil {
		return fmt.Errorf("reviewer: create deny check: %w", err)
	}
	return nil
}

// alreadyPublished reports whether the agent-review check already exists for the head SHA — i.e.
// this SHA was already published and a redelivered task should not re-post. A lookup error is
// treated as "not published" so a transient failure never suppresses a real review.
func (e *Engine) alreadyPublished(ctx context.Context, meta publishMeta) bool {
	res, err := e.gh.AgentCheck(ctx, meta.owner, meta.repo, meta.headSHA, checkName)
	return err == nil && res.Found
}

// classify splits confidence-gated findings into inline findings (actionable, on a commentable diff
// line — these become review comments after reconciliation), out-of-diff actionable findings (listed
// in the summary, never snapped to a wrong line — spec Decision 6), and nitpicks (collapsed in the
// summary).
func classify(findings []finding, idx diffIndex) (inline, outOfDiff, nitpicks []finding) {
	for _, f := range findings {
		if f.Severity == SeverityNitpick {
			nitpicks = append(nitpicks, f)
			continue
		}
		if f.File != "" && f.Line > 0 && idx.inDiff(f.File, f.Line) {
			inline = append(inline, f)
			continue
		}
		outOfDiff = append(outOfDiff, f)
	}
	return inline, outOfDiff, nitpicks
}

// inlineCommentBody renders one inline comment: an icon+category prefix, the message, an optional
// plain-language suggestion, and an optional "Prompt for AI agents" block (spec Decisions 9/10).
//
// The suggestion is prose, deliberately not a GitHub ```suggestion block. That block is a
// one-click commit of a verbatim replacement for the commented lines, and the lenses cannot
// author one reliably: they see a unified diff with no surrounding file, so a snippet they
// write rarely aligns with the exact lines it would replace, and "apply" then commits broken
// code. Worse, the model-authored "snippet" is often itself a sentence, which the block would
// offer to commit as source. A sentence describing the change is what the reader can act on
// with the context they have; the fenced fix prompt below is the hand-off for an agent that
// does have the file.
func inlineCommentBody(f finding) string {
	var b strings.Builder
	// Dimension/severity are normalized to known enums, so only the model-authored message needs
	// sanitizing here.
	fmt.Fprintf(&b, "**%s** · _%s_\n\n%s\n", findingPrefix(f), f.Dimension, sanitizeText(f.Message))
	if s := strings.TrimSpace(f.Suggestion); s != "" {
		// Model-authored prose, so it gets the same sanitizing as the message.
		fmt.Fprintf(&b, "\n**Suggestion:** %s\n", sanitizeText(s))
	}
	if f.FixPrompt != "" {
		// FixPrompt is model-authored; render it inside a code fence so any @mentions or HTML are
		// literal (not pinged/injected) and it stays copy-pasteable. Size the fence past any
		// backtick run in the prompt so the content cannot break out.
		fence := strings.Repeat("`", maxBacktickRun(f.FixPrompt)+1)
		if len(fence) < 3 {
			fence = "```"
		}
		b.WriteString("\n<details>\n<summary>🤖 Prompt for AI agents</summary>\n\n")
		b.WriteString(fence + "\n")
		b.WriteString(f.FixPrompt)
		if !strings.HasSuffix(f.FixPrompt, "\n") {
			b.WriteByte('\n')
		}
		b.WriteString(fence + "\n\n</details>\n")
	}
	// Hidden fingerprint marker so a later re-review re-identifies this comment and reconciles it
	// (GitHub-as-store); it renders invisibly.
	b.WriteString("\n" + fpMarker(f.fingerprint()) + "\n")
	return b.String()
}

// sanitizeText neutralizes model-authored text for safe embedding in a Markdown comment: it
// escapes HTML-significant characters (so a finding can't inject markup such as </details>) and
// breaks @mentions with a zero-width space (so the reviewer never pings a real user). The fenced
// FixPrompt is left untouched by callers — a code fence already renders its content literally.
func sanitizeText(s string) string {
	s = strings.ReplaceAll(s, "&", "&amp;")
	s = strings.ReplaceAll(s, "<", "&lt;")
	s = strings.ReplaceAll(s, ">", "&gt;")
	return mentionPattern.ReplaceAllString(s, "@\u200b$1")
}

// mentionPattern matches an @ immediately followed by a mention character; sanitizeText inserts a
// zero-width space after the @ so GitHub does not render (and notify) it as a mention.
var mentionPattern = regexp.MustCompile(`@([A-Za-z0-9])`)

// findingPrefix is the icon+category label that leads an inline comment (spec Decision 9).
func findingPrefix(f finding) string {
	if f.Dimension == DimSecurity {
		return "🔒 Security"
	}
	switch f.Severity {
	case SeverityCritical, SeverityMajor:
		return "⚠️ Potential issue"
	default:
		return "🛠️ Refactor"
	}
}

// summaryComment assembles the marker-updated summary comment (spec Decision 9): header, scorecard
// table, and collapsible sections for nitpicks, out-of-diff findings, and review details.
func summaryComment(marker string, card scorecard, actionable int, nitpicks, outOfDiff []finding, meta publishMeta) string {
	var b strings.Builder
	b.WriteString(marker)
	b.WriteByte('\n')
	fmt.Fprintf(&b, "## %s Agent review — Overall: %s · Actionable comments: %d\n\n", card.overall, levelWord(card.overall), actionable)
	b.WriteString(scorecardTable(card))
	if len(nitpicks) > 0 {
		b.WriteString(collapsible(fmt.Sprintf("🧹 Nitpicks (%d)", len(nitpicks)), findingsList(nitpicks)))
	}
	if len(outOfDiff) > 0 {
		b.WriteString(collapsible(fmt.Sprintf("🔭 Outside diff range (%d)", len(outOfDiff)), findingsList(outOfDiff)))
	}
	b.WriteString(collapsible("Review details", reviewDetails(card, meta)))
	return b.String()
}

// scorecardTable renders the per-dimension severity histogram (spec Decision 5). With no findings
// it states so rather than emitting an empty table.
func scorecardTable(card scorecard) string {
	if len(card.dims) == 0 {
		return "_No findings._\n\n"
	}
	var b strings.Builder
	b.WriteString("| Dimension | Level | Critical | Major | Medium | Nitpick |\n")
	b.WriteString("|---|---|---|---|---|---|\n")
	for _, d := range card.dims {
		fmt.Fprintf(&b, "| %s | %s | %d | %d | %d | %d |\n", d.dimension, d.level, d.critical, d.major, d.medium, d.nitpick)
	}
	b.WriteByte('\n')
	return b.String()
}

// findingsList renders findings as a bulleted file:line list for the summary's collapsible
// sections (out-of-diff findings and nitpicks).
func findingsList(fs []finding) string {
	var b strings.Builder
	for _, f := range fs {
		loc := f.File
		if f.Line > 0 {
			loc = fmt.Sprintf("%s:%d", f.File, f.Line)
		}
		fmt.Fprintf(&b, "- **%s** `%s` _(%s)_ — %s\n", f.Severity, loc, f.Dimension, sanitizeText(f.Message))
	}
	return b.String()
}

// reviewDetails renders the "Review details" section: head SHA, file count, the standards
// applied, and the per-lens status table.
func reviewDetails(card scorecard, meta publishMeta) string {
	var b strings.Builder
	fmt.Fprintf(&b, "- Head SHA: `%s`\n", meta.headSHA)
	fmt.Fprintf(&b, "- Files reviewed: %d\n", len(meta.files))
	if len(meta.standards) > 0 {
		fmt.Fprintf(&b, "- Standards applied: %s\n", strings.Join(meta.standards, ", "))
	} else {
		// Empty also covers standards-off and the discovery/distill fallback, not just a repo with
		// no convention docs — so stay neutral rather than asserting none were found.
		b.WriteString("- Standards: generic review\n")
	}
	if len(meta.lenses) > 0 {
		b.WriteByte('\n')
		b.WriteString(lensTable(card, meta.lenses))
	}
	return b.String()
}

// lensTable renders one status row per lens — every category and the glue pass, whether or not
// it ran, unlike the scorecard, which lists only dimensions with findings: its level (derived
// from the scorecard, see lensLevel), the model it ran on, how long it took, and the tokens it
// consumed. A lens the diff did not select is marked skipped, and one that produced no output is
// marked so a silent lens is visible rather than reading as a clean one; token cells show "–"
// when the model reported no usage, which is not the same as zero.
func lensTable(card scorecard, lenses []lensStat) string {
	var b strings.Builder
	b.WriteString("| Lens | Level | Model | Time | Tokens in | Tokens out |\n")
	b.WriteString("|---|---|---|---|---|---|\n")
	for _, l := range lenses {
		if l.skipped {
			fmt.Fprintf(&b, "| %s | ⏭️ skipped (no UI files changed) | – | – | – | – |\n", l.lens.title)
			continue
		}
		lvl := lensLevel(card, l.lens).String()
		if !l.ran {
			lvl = "⚪ no output"
		}
		in, out := "–", "–"
		if l.usage {
			in, out = fmt.Sprint(l.tokensIn), fmt.Sprint(l.tokensOut)
		}
		fmt.Fprintf(&b, "| %s | %s | `%s` | %s | %s | %s |\n", l.lens.title, lvl, l.model, formatElapsed(l.elapsed), in, out)
	}
	return b.String()
}

// formatElapsed renders a lens duration for the table as seconds to one decimal — "0.9s",
// "12.3s", "74.5s" — one unit down the column so rows compare at a glance; "–" when nothing
// was observed.
func formatElapsed(d time.Duration) string {
	if d <= 0 {
		return "–"
	}
	return fmt.Sprintf("%.1fs", d.Seconds())
}

// collapsible wraps body in a <details> block with the given summary label.
func collapsible(summary, body string) string {
	return fmt.Sprintf("\n<details>\n<summary>%s</summary>\n\n%s\n</details>\n", summary, body)
}

// levelWord is the textual grade shown beside the glyph in headers and the check.
func levelWord(l level) string {
	switch l {
	case levelRed:
		return "Red"
	case levelYellow:
		return "Yellow"
	default:
		return "Green"
	}
}

// checkConclusion maps the overall grade to the advisory check conclusion (spec Decision 15):
// green is success; yellow and red are neutral. It is never failure — the reviewer never gates a
// merge.
func checkConclusion(overall level) string {
	if overall == levelGreen {
		return "success"
	}
	return "neutral"
}

File: kotlin/src/main/kotlin/com/automation/agent/agent/reviewer/Publish.kt
/*
 * The CodeRabbit-style publish stage: assembly + REST writes (advisory review, marker summary
 * comment, advisory agent-review check), reconciled against the PR's existing fingerprinted
 * comments. Nothing here gates a merge.
 */
package com.automation.agent.agent.reviewer

import com.automation.agent.githubapi.CheckRunInput
import com.automation.agent.githubapi.PRFile
import com.automation.agent.githubapi.ReviewComment
import com.automation.agent.githubapi.ReviewInput


// The advisory check the reviewer publishes (agent-published, human-consumed). Globally unique and
// identical across ports (external contract).
const val CHECK_NAME = "agent-review"

/** Carries the per-PR identifiers and context the published artifacts need. */
data class PublishMeta(
    val owner: String,
    val repo: String,
    val number: Int,
    val headSha: String,
    val files: List<PRFile> = emptyList(), // for the in-diff index
    val tiers: String = "", // model tiers used, for the Review details section
    var standards: List<String> = emptyList(), // applied source paths (empty = generic)
)

/**
 * The hidden HTML comment that identifies the reviewer's single summary comment so a re-review
 * updates it rather than posting a new one.
 */
fun summaryMarker(owner: String, repo: String, number: Int): String = "<!-- automation-agent:review:$owner/$repo#$number -->"

/**
 * Posts the review for a scored PR: inline comments for in-diff actionable findings, a
 * marker-updated summary comment with the scorecard, and the advisory agent-review check.
 * Out-of-diff actionable findings and nitpicks go into the summary (never dropped).
 */
suspend fun publish(engine: Engine, card: Scorecard, findings: List<Finding>, meta: PublishMeta) {
    val gh = engine.requireGh()
    // At-least-once safety: reconciliation makes the inline comments idempotent, but the check run and
    // summary are create/upsert-only, so a redelivered task for a SHA already published would
    // duplicate the check. If the agent-review check already exists for this head SHA, skip — a
    // genuine re-push carries a new SHA and still reconciles below.
    if (alreadyPublished(engine, meta)) {
        engine.log.log(System.Logger.Level.INFO, "review already published for head SHA; skipping re-post repo=${meta.owner}/${meta.repo} sha=${meta.headSha}")
        return
    }
    val idx = DiffIndex(meta.files)
    val (inline, outOfDiff, nitpicks) = classify(findings, idx)
    val actionable = inline.size + outOfDiff.size

    // Reconcile against the comments already on the PR (GitHub-as-store): keep inline findings that
    // still apply (don't re-post — idempotent), post only new ones, and minimize the comments whose
    // finding is gone.
    val existing = gh.listReviewComments(meta.owner, meta.repo, meta.number)
    val rec = reconcile(inline, existing)

    // Post only the new inline findings; an empty review is noise.
    if (rec.toPost.isNotEmpty()) {
        val comments = rec.toPost.map { f -> ReviewComment(path = f.file, line = f.line, side = "RIGHT", body = inlineCommentBody(f)) }
        val body = "${levelGlyph(card.overall)} Agent review — see the summary comment for the full scorecard."
        gh.createReview(meta.owner, meta.repo, meta.number, ReviewInput(body = body, comments = comments))
    }

    // Minimize the comments whose finding no longer applies — best-effort. New inline comments are
    // already posted but the summary and check are not; aborting here on a single minimize failure
    // would leave the PR without its summary/check. So log and continue per node.
    for (nodeId in rec.toMinimize) {
        try {
            gh.minimizeComment(nodeId)
        } catch (e: Exception) {
            engine.log.log(System.Logger.Level.WARNING, "reviewer: minimize outdated comment failed; continuing repo=${meta.owner}/${meta.repo} node=$nodeId err=${e.message}")
        }
    }

    val marker = summaryMarker(meta.owner, meta.repo, meta.number)
    gh.upsertMarkerComment(meta.owner, meta.repo, meta.number, marker, summaryComment(marker, card, actionable, nitpicks, outOfDiff, meta))

    gh.createCheckRun(
        meta.owner,
        meta.repo,
        CheckRunInput(
            name = CHECK_NAME,
            headSha = meta.headSha,
            conclusion = checkConclusion(card.overall),
            title = "${levelGlyph(card.overall)} Agent review — ${levelWord(card.overall)}",
            summary = "Overall: ${levelWord(card.overall)} · Actionable comments: $actionable",
        ),
    )
}

/**
 * Posts the "too large to review" outcome: a marker-updated summary comment framed fail-like (🔴)
 * plus a neutral check. No model call was made.
 */
suspend fun publishDeny(engine: Engine, meta: PublishMeta, reason: String, files: Int, diffBytes: Int) {
    val gh = engine.requireGh()
    if (alreadyPublished(engine, meta)) {
        engine.log.log(System.Logger.Level.INFO, "deny already published for head SHA; skipping re-post repo=${meta.owner}/${meta.repo} sha=${meta.headSha}")
        return
    }
    val marker = summaryMarker(meta.owner, meta.repo, meta.number)
    val body =
        "$marker\n## 🔴 Agent review — too large for automated review\n\n" +
            "This PR is too large to review automatically ($files files / $diffBytes bytes " +
            "after excluding generated files). Please split it into smaller PRs.\n\n" +
            "_${reason}_\n"
    gh.upsertMarkerComment(meta.owner, meta.repo, meta.number, marker, body)
    gh.createCheckRun(
        meta.owner,
        meta.repo,
        CheckRunInput(
            name = CHECK_NAME,
            headSha = meta.headSha,
            conclusion = "neutral",
            title = "🔴 Agent review — too large",
            summary = "$files files / $diffBytes bytes after excluding generated files; please split.",
        ),
    )
}

/**
 * Reports whether the agent-review check already exists for the head SHA. A lookup error is treated
 * as "not published" so a transient failure never suppresses a real review.
 */
suspend fun alreadyPublished(engine: Engine, meta: PublishMeta): Boolean =
    try {
        engine.requireGh().agentCheck(meta.owner, meta.repo, meta.headSha, CHECK_NAME).found
    } catch (_: Exception) {
        false
    }

/** The three-way split of confidence-gated findings the publish stage routes on. */
data class Classified(val inline: List<Finding>, val outOfDiff: List<Finding>, val nitpicks: List<Finding>)

/**
 * Splits confidence-gated findings into inline findings (actionable, on a commentable diff line),
 * out-of-diff actionable findings (listed in the summary, never snapped to a wrong line), and
 * nitpicks (collapsed in the summary).
 */
fun classify(findings: List<Finding>, idx: DiffIndex): Classified {
    val inline = mutableListOf<Finding>()
    val outOfDiff = mutableListOf<Finding>()
    val nitpicks = mutableListOf<Finding>()
    for (f in findings) {
        when {
            f.severity == Severity.NITPICK -> nitpicks.add(f)
            f.file != "" && f.line > 0 && idx.inDiff(f.file, f.line) -> inline.add(f)
            else -> outOfDiff.add(f)
        }
    }
    return Classified(inline = inline, outOfDiff = outOfDiff, nitpicks = nitpicks)
}

/**
 * Renders one inline comment: an icon+category prefix, the message, an optional ```suggestion block
 * (a localized fix), and an optional "Prompt for AI agents" block.
 */
fun inlineCommentBody(f: Finding): String {
    val parts = StringBuilder()
    // Dimension/severity are normalized to known enums, so only the model-authored message needs
    // sanitizing here.
    parts.append("**${findingPrefix(f)}** · _${f.dimension.value}_\n\n${sanitizeText(f.message)}\n")
    if (f.suggestion != "") {
        // Suggestion is model-authored; size the outer fence past any backtick run in it so a suggestion
        // containing a ```fence can't close the block early and inject markdown or @mentions.
        var fence = "`".repeat(maxBacktickRun(f.suggestion) + 1)
        if (fence.length < 3) fence = "```"
        parts.append("\n").append(fence).append("suggestion\n")
        parts.append(f.suggestion)
        if (!f.suggestion.endsWith("\n")) parts.append("\n")
        parts.append(fence).append("\n")
    }
    if (f.fixPrompt != "") {
        // FixPrompt is model-authored; render it inside a code fence so any @mentions or HTML are
        // literal (not pinged/injected) and it stays copy-pasteable.
        var fence = "`".repeat(maxBacktickRun(f.fixPrompt) + 1)
        if (fence.length < 3) fence = "```"
        parts.append("\n<details>\n<summary>🤖 Prompt for AI agents</summary>\n\n")
        parts.append(fence).append("\n")
        parts.append(f.fixPrompt)
        if (!f.fixPrompt.endsWith("\n")) parts.append("\n")
        parts.append(fence).append("\n\n</details>\n")
    }
    // Hidden fingerprint marker so a later re-review re-identifies this comment and reconciles it.
    parts.append("\n").append(fpMarker(fingerprint(f))).append("\n")
    return parts.toString()
}

// Matches an @ immediately followed by a mention character; sanitizeText inserts a zero-width space
// after the @ so GitHub does not render (and notify) it as a mention.
private val MENTION_PATTERN = Regex("@([A-Za-z0-9])")

/**
 * Neutralizes model-authored text for safe embedding in a Markdown comment: escape HTML-significant
 * characters (so a finding can't inject markup such as </details>) and break @mentions with a
 * zero-width space (so the reviewer never pings a real user). Code in ```suggestion blocks and
 * fenced FixPrompt is left untouched by callers.
 */
fun sanitizeText(s: String): String {
    var out = s.replace("&", "&amp;")
    out = out.replace("<", "&lt;")
    out = out.replace(">", "&gt;")
    return MENTION_PATTERN.replace(out, "@​$1")
}

/** The icon+category label that leads an inline comment. */
fun findingPrefix(f: Finding): String {
    if (f.dimension == Dimension.SECURITY) return "🔒 Security"
    if (f.severity == Severity.CRITICAL || f.severity == Severity.MAJOR) return "⚠️ Potential issue"
    return "🛠️ Refactor"
}

/**
 * Assembles the marker-updated summary comment: header, scorecard table, and collapsible sections
 * for nitpicks, out-of-diff findings, and review details.
 */
fun summaryComment(marker: String, card: Scorecard, actionable: Int, nitpicks: List<Finding>, outOfDiff: List<Finding>, meta: PublishMeta): String {
    val parts = StringBuilder()
    parts.append(marker).append("\n")
    parts.append("## ${levelGlyph(card.overall)} Agent review — Overall: ${levelWord(card.overall)} · Actionable comments: $actionable\n\n")
    parts.append(scorecardTable(card))
    if (nitpicks.isNotEmpty()) parts.append(collapsible("🧹 Nitpicks (${nitpicks.size})", findingsList(nitpicks)))
    if (outOfDiff.isNotEmpty()) parts.append(collapsible("🔭 Outside diff range (${outOfDiff.size})", findingsList(outOfDiff)))
    parts.append(collapsible("Review details", reviewDetails(meta)))
    return parts.toString()
}

/**
 * Renders the per-dimension severity histogram. With no findings it states so rather than emitting
 * an empty table.
 */
fun scorecardTable(card: Scorecard): String {
    if (card.dims.isEmpty()) return "_No findings._\n\n"
    val parts = StringBuilder()
    parts.append("| Dimension | Level | Critical | Major | Medium | Nitpick |\n")
    parts.append("|---|---|---|---|---|---|\n")
    for (d in card.dims) {
        parts.append("| ${d.dimension.value} | ${levelGlyph(d.level)} | ${d.critical} | ${d.major} | ${d.medium} | ${d.nitpick} |\n")
    }
    parts.append("\n")
    return parts.toString()
}

/** Renders findings as a bulleted file:line list for the summary's collapsible sections. */
fun findingsList(fs: List<Finding>): String {
    val parts = StringBuilder()
    for (f in fs) {
        val loc = if (f.line > 0) "${f.file}:${f.line}" else f.file
        parts.append("- **${f.severity.value}** `$loc` _(${f.dimension.value})_ — ${sanitizeText(f.message)}\n")
    }
    return parts.toString()
}

/** Renders the "Review details" section: head SHA, file count, and the model tiers. */
fun reviewDetails(meta: PublishMeta): String {
    val parts = StringBuilder()
    parts.append("- Head SHA: `${meta.headSha}`\n")
    parts.append("- Files reviewed: ${meta.files.size}\n")
    if (meta.tiers != "") parts.append("- Model tiers: ${meta.tiers}\n")
    if (meta.standards.isNotEmpty()) {
        parts.append("- Standards applied: ${meta.standards.joinToString(", ")}\n")
    } else {
        // Empty also covers standards-off and the discovery/distill fallback, not just a repo with no
        // convention docs — so stay neutral rather than asserting none were found.
        parts.append("- Standards: generic review\n")
    }
    return parts.toString()
}

/** Wraps [body] in a <details> block with the given summary label. */
fun collapsible(summary: String, body: String): String = "\n<details>\n<summary>$summary</summary>\n\n$body\n</details>\n"

/**
 * Maps the overall grade to the advisory check conclusion: green is success; yellow and red are
 * neutral. It is never failure — the reviewer never gates a merge.
 */
fun checkConclusion(overall: Level): String = if (overall == Level.GREEN) "success" else "neutral"
